codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/Paginators.java

```diff
@@ -0,0 +1,254 @@
+        writer.openBlock("func (p $P) NextPage(ctx $T, optFns ...func(*Options)) ($P, error) {", "}",
+                paginatorSymbol, contextSymbol, outputSymbol, () -> {
+                    writer.addUseImports(SmithyGoDependency.FMT);
+                    writer.openBlock("if !p.HasMorePages() {", "}", () -> {
+                        writer.write("return nil, fmt.Errorf(\"no more pages available\")");
+                    });
+                    writer.write("");
+                    writer.write("params := *p.params");
+                    writer.write("params.$L = p.nextToken", inputMember);
+                    pageSizeMember.ifPresent(memberShape -> writer.write("params.$L = p.options.Limit",
+                            symbolProvider.toMemberName(memberShape)));
+
+                    writer.write("result, err := p.client.$L(ctx, &params, optFns...)",
+                            operationSymbol.getName());
+                    writer.openBlock("if err != nil {", "}", () -> {
+                        writer.write("return nil, err");
+                    });
+                    writer.write("p.firstPage = false");
+
+                    StringBuilder nilGuard = new StringBuilder();
+                    StringBuilder outputPath = new StringBuilder("result");
+
+                    List<MemberShape> outputMemberPath = paginationInfo.getOutputTokenPath();
+                    for (int i = 0; i < outputMemberPath.size(); i++) {
+                        MemberShape memberShape = outputMemberPath.get(i);
+                        outputPath.append(".");
+                        outputPath.append(symbolProvider.toMemberName(memberShape));
+
+                        // Don't need to check the last member here as we won't dereference it
+                        if (i != outputMemberPath.size() - 1) {
+                            if (i != 0) {
+                                nilGuard.append(" && ");
+                            }
+                            nilGuard.append(outputPath);
+                            nilGuard.append(" != nil");
+                        }
+                    }
+
+                    Runnable setToken = () -> {
+                        writer.write("p.nextToken = $L", outputPath);
+                    };
+                    if (nilGuard.length() > 0) {
+                        writer.openBlock("if $L {", "}", nilGuard.toString(), setToken::run);
+                    } else {
+                        setToken.run();
+                    }
+
+                    writer.openBlock("if p.options.StopOnDuplicateToken && "
+                            + "params.$L != nil && p.nextToken != nil && "
+                            + "*params.$L == *p.nextToken {", "", inputMember, inputMember, () -> {
+                        writer.write("p.done = true");
+                        writer.openBlock("} else {", "}", () -> writer.write("p.done = false"));
```

Comment:
If a paginator stops because of duplicate tokens and user asked it to stop, this should be considered terminal stop.  Probably no need to allow resume. Doing so seems to be a recipe for unexpected behavior and errors.